task-canvas/script.js

```diff
@@ -0,0 +1,490 @@
+
+        return path.join(' ');
+    }
+
+    render() {
+        if (this.coords.length == 1) {
+            this.coords = this.coords.flat();
+        }
+        this.drawShape(this.coords[0], this.coords[1], this.coords[2], this.coords[3]);
+    }
+
+    drawShape(x, y, vx, vy) {
+        let canvas = document.getElementById('canvas');
+        let ctx = canvas.getContext('2d');
+        let radius = Math.sqrt((vx - x) ** 2 + (vy - y) ** 2);
+        ctx.strokeStyle = this.color;
+        ctx.beginPath();
+        ctx.arc(x, y, radius, 0, Math.PI * 2, false);
+        ctx.stroke();
+    }
+
+    isPointWithin(x, y) {
+        let t = 10;
+        if (Math.abs(Math.sqrt((x - this.coords[0])**2 + (y - this.coords[1])**2) - Math.sqrt((this.coords[2] - this.coords[0]) ** 2 + (this.coords[3] - this.coords[1]) ** 2)) <= t) {
+            return true;
+        }
+        return false;
+    }
+
+    setCoord(x, y) {
+        for (let i = 0; i < (this.coords.length); i += 2) {
+            this.coords[i] += x;
+        }
+        for (let j = 1; j < (this.coords.length); j += 2) {
+            this.coords[j] += y;
+        }
+    }
+
+    setColor(color) {
+        this.color = color;
+    }
+}
+
+class CanvasVectorEditor {
+    constructor() {
+        this.canvas = document.getElementById('canvas');
+        this.colorPicker = document.getElementById('colors');
+        this.color = '#08D9D6';
+        this.coords = {
+            x: 0,
+            y: 0,
+            vx: 0,
+            vy: 0,
+        };
+        this.amountCoords = [];
+        this.arrayOfShapes = [];
+        this.movingShape;
+        this.move = this.mouseMove.bind(this);
+        this.remove = document.getElementById('remove');
+        this.savingPng = document.getElementById('savingAsPng');
+        this.savingSVG = document.getElementById('savingAsSvg');
+        this.impJson = document.getElementById('importJSON');
+        this.expJson = document.getElementById('exportJSON');
+        this.initializeSubscriptions();
+    }
+
+    initializeSubscriptions() {
+        this.canvas.addEventListener('mousedown', this.mouseDown.bind(this));
+        this.canvas.addEventListener('mouseup', () => { this.finishShape() }, false );
+        this.colorPicker.addEventListener("input", this.newColor.bind(this), false);
+        this.colorPicker.addEventListener("change", this.watchColorPicker.bind(this), false);
+        this.remove.addEventListener("click", this.removeAll.bind(this), false);
+        window.addEventListener('beforeunload', this.closeWindowAndTab.bind(this));
+        window.addEventListener('load', this.loadWindow.bind(this));
+        this.savingPng.addEventListener('click', this.saveImagePng.bind(this));
+        this.impJson.addEventListener('click', this.importjson.bind(this));
+        this.expJson.addEventListener('click', this.exporjson.bind(this));
+        this.savingSVG.addEventListener('click', this.saveImageSvg.bind(this));
```

Comment:
In most cases `.bind(this)` is an obsolete thing from ES5 standard.
So, for example, it's better to `this.savingSVG.addEventListener('click', this.saveImageSvg);` plus define `saveImageSvg` as class property + arrow function
`saveImageSvg = () => { ... };`
(the same for other methods with `.bind(this)`)

Reply:
I din't understand about class property.
`bind(this)` I used only in functions with argument event.